charmhelpers/core/host.py

```diff
@@ -756,6 +757,25 @@ def restart_on_change_helper(lambda_f, restart_map, stopstart=False,
     return r
 
 
+def restart_pid_check(service_name, ptable_string=None):
+    """Stop a service, check the processes are gone, start service
+    @param service_name: service name as init system knows it
+    @param ptable_string: string to look for in process table to match service
+    """
+
+    @retry_on_exception(5, base_delay=3, exc_type=AssertionError)
+    def check_pids_gone(svc_string):
+        log("Checking no pids for {} exist".format(svc_string), level=INFO)
+        assert(subprocess.call(["pgrep", svc_string, "--nslist", "pid",
+                               "--ns", str(os.getpid())]) == 1)
+
+    if not ptable_string:
+        ptable_string = service_name
+    service_stop(service_name)
+    check_pids_gone(ptable_string)
```

Comment:
So if this fails, then the function will throw an `AssertionException` which  I doubt will be caught, and the hook will crash.  Is this really what is wanted at this stage?

This function is supposed to restart the associated `service_name`, but stopping it, making sure the associated processes are gone, and then restarting it.  But it gives up!

Would it not be better to catch the AssertionError and then just `pkill` those processes and wait for them to die?  i.e. let's be nice, but if they won't play ball, kill them.  And then restart the service.  This way the intent of the function is preserved *and* it's what the user would have to do anyway.